llama_index/callbacks/token_counting.py

```diff
@@ -0,0 +1,147 @@
+            (see llama_index.utils.globals_helper).
+        event_starts_to_ignore: List of event types to ignore at the start of a trace.
+        event_ends_to_ignore: List of event types to ignore at the end of a trace.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Optional[Callable[[str], List]] = None,
+        event_starts_to_ignore: Optional[List[CBEventType]] = None,
+        event_ends_to_ignore: Optional[List[CBEventType]] = None,
+        verbose: bool = False,
+    ) -> None:
+        self.llm_token_counts: List[TokenCountingEvent] = []
+        self.embedding_token_counts: List[TokenCountingEvent] = []
+        self.tokenizer = tokenizer or globals_helper.tokenizer
+        self._verbose = verbose
+
+        super().__init__(
+            event_starts_to_ignore=event_starts_to_ignore or [],
+            event_ends_to_ignore=event_ends_to_ignore or [],
+        )
+
+    def start_trace(self, trace_id: Optional[str] = None) -> None:
+        return
+
+    def end_trace(
+        self,
+        trace_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        return
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> None:
+        """Count the LLM or Embedding tokens as needed."""
+        if (
+            event_type == CBEventType.LLM
+            and event_type not in self.event_ends_to_ignore
+            and payload is not None
+        ):
+            self.llm_token_counts.append(
+                TokenCountingEvent(
+                    event_id=event_id,
+                    prompt=payload.get("formatted_prompt", ""),
+                    prompt_token_count=len(
+                        self.tokenizer(payload.get("formatted_prompt", ""))
+                    ),
+                    completion=payload.get("response", ""),
+                    completion_token_count=len(
+                        self.tokenizer(payload.get("response", ""))
+                    ),
+                )
+            )
+
+            if self._verbose:
+                print(
+                    "LLM Prompt Token Usage: "
+                    f"{self.llm_token_counts[-1].prompt_token_count}\n"
+                    "LLM Completion Token Usage: "
+                    f"{self.llm_token_counts[-1].completion_token_count}",
+                    flush=True,
+                )
+        elif (
+            event_type == CBEventType.EMBEDDING
+            and event_type not in self.event_ends_to_ignore
+            and payload is not None
+        ):
+            total_chunk_tokens = 0
+            for chunk in payload.get("chunks", []):
```

Comment:
So the assumption here is that the event producers publish events with this particular payload?

I think at minimum, we should use a constant for the key here. So that it's clear that there's this dependency.